Make imagelib MirageOS-friendly

cfcs opened the pull request:
Hello!

I really wanted PNG rendering in MirageOS; I got that working with this patch (so now I have a portable "gallery viewer" that works on Linux and Xen/Qubes - very happy!).

It's a bit invasive, sorry about that. If there are things you'd like to change, I'd be happy to do that. :-)

I only tested PNG parsing, so there may be subtle bugs introduced. I hope not, but now you are warned.

This patch:
- switches the compression library from `camlzip` to the pure `decompress` (addressing #1 )
- changes the name `openfile` to `parsefile` (hope that is OK)
- replaces `String.lowercase` with `String.lowercase_ascii` (to get rid of deprecation warnings)
- adds `safe_string` (which will be default in future OCaml releases, hopefully soon), enforcing the distinction between immutable strings (`string`) and mutable strings (`Bytes.t`)
- changes the chunk reading functions from the Unix/Windows-only `in_channel` functions to use a pluggable method, defined at the top of `ImageUtil.ml`
  - implements `ImageUtil.chunk_reader_of_in_channel` and `chunk_reader_of_path` for backwards compatibility, so `ImageLib.openfile` still works.
- in `imagePNG.ml`, which was the file I mainly worked on, it also
  - switches `Scanf.fscanf`, which is deprecated since 4.03.0 due to bugs, to `Scanf.bscanf` (`fscanf` also doesn't work on MirageOS)

ping @rlepigre @dinosaure

Commit:
make imagelib mirage-friendly

Commit:
remove camlzip dependency and switch to the decompress library
- (and fix some errors introduced in last commit)
- also add .merlin file

rlepigre commented:
Cool! Thanks for the contribution!

Since you are offering :-), here are a few thinks I'd like you to change before I merge your pull-request:
- I'm not such a big fan of the result type thing (and monadic programing style in general), so it would be great if you could at least go back to the old "imageLib.mli" interface (no result type for "openfile").
- I ran some tests, and it seems that there is a problem with the PPM/PGM/PBM format. I actually ended up with the same kind of problem when I tried to remove the calls to "fscanf"... I'm sure it's nothing, but I did not have time to debug this stuff. Do you have any idea what the problem might be?

Commit:
ImageLib.openfile: return to exception-based error reporting

cfcs commented:
Oh, ok, I like that :) I changed the `ImageLib.openfile` interface back -- do you think this is acceptable, or should I get rid of the result type altogether?

Re: PPM: Yes, the code did `seek_in 0` before and read the header two times. I removed that to make it fit with the current code. This causes it to ignore the first byte of that PPM contents. I'll try to fix it and report back.

Commit:
imagePPM: re-use Scanf.Scanner from read_header

rlepigre commented:
I'm not bothered as much by the remaining uses of the result type. You can leave them if you want, but it is even better if you remove them! ;)

For the PPM problem, I completely missed the seek_in think, I completely forgot about it... Actually, I can't even believe that this is my own code! Anyway, thanks a lot for looking into this!

cfcs commented:
@rlepigre I made it re-use the scanner from `read_header`, does this work for you?

rlepigre commented:
No, still does not work (at least with binary version of the PPM format)... You can test with the attached program and examples.
[imagelib_test.tar.gz](https://github.com/rlepigre/ocaml-imagelib/files/1424329/imagelib_test.tar.gz)

rlepigre commented:
I'm thinking that I should set up a test suite, I'm gonna do that some day... :-)

Actually, when I originally developed imagelib, I used a Patoline document to run tests, so I had a PDF file with all my test images inside. But this is not helpful when there is an error.

cfcs commented:
Thanks, I will look into it!

cfcs commented:
Ah, sorry, I'm a bit tired today. The module contains 6 parsers. The text-based file I was testing with was a `P2` I believe, which means it uses `bscanf`.
`P4`, `P5`, `P6` read bytes directly. So same issue as before. Will be back with a fix.

Commit:
imagePPM: implement buffer for one-character reads to enable mixed chunk_byte / Scanf.bscanf calls

cfcs commented:
@rlepigre This is a bit dirty, but it does make PPM parsing work for your testcases (and now that I sat down and read the entire function, I think it will work for all). It's a little bit hackish, but I didn't see another easy way to do it without rewriting everything to either use `Scanf` or not.

rlepigre commented:
OK, let's merge it then!

rlepigre commented:
I added you as a collaborator so you can push to the repository directly! :)

cfcs commented:
Cool! I saw that I missed some text in the README, added another PR for that. :)
